### PR TITLE
Upgrade to Rails 6.1

### PR DESCRIPTION
**Pull request by malcolmbaig**
## Context
Where possible we should be using the latest version of Rails.

## Changes proposed in this pull request
- Fix errors found by our test suite
- Fix several (but not all) deprecation warnings

## Guidance to review
- Per-commit recommended
- A quick click around the review app to ensure things are working as expected would be helpful
- There are still a few deprecation warnings left, mostly to do with changes to the ActiveModel::Errors API. There's probably enough going on in this PR, will fix remaining warnings in a followup.
- Can anyone figure out why 4940f222f1ee61a59c11ee076df9cade0dada606 is necessary?

## Link to Trello card
https://trello.com/c/HCpYZuT0

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: config/application.rb [hunk 1]
```diff
@@ -52,7 +52,7 @@ class Application < Rails::Application
     config.action_view.default_form_builder = GOVUKDesignSystemFormBuilder::FormBuilder
 
     config.i18n.exception_handler = Proc.new { |exception| raise exception.to_exception }
-    config.action_view.raise_on_missing_translations = true
+    config.i18n.raise_on_missing_translations = true
```
Comment (duncanjbrown) [inline config/application.rb:55]: Does this make https://github.com/DFE-Digital/apply-for-teacher-training/pull/4220 redundant?

**Review by duncanjbrown — APPROVED** [review]
Thanks @malcolmbaig!  Nice commits too 💎. Review app seems ok to me.

File: config/application.rb (hunk 1 above)
Comment (malcolmbaig) [reply config/application.rb:55]: Unfortunately still need the custom exception handler to also raise errors in the case of activemodel errors that don't resolve correctly and don't have a default (eg — errors for custom validations).

**Comment by malcolmbaig** [comment]
Updated the commit message for https://github.com/DFE-Digital/apply-for-teacher-training/pull/4486/commits/b3c11b0f240ea6726b90db45637b3051b6d97131

**Pull request merged**